Fix theme overwrite for subdir deployments

phisch opened the pull request:
## Description
Because of a confusion with `webroot`, `serverroot`, `relative/absolute` paths and a mixture of adding and removing trailing and leading slashes, an app theme could not overwrite resources when the owncloud instance was configured like described in #27630. This change fixes this.

## Related Issue
#27630


## How Has This Been Tested?
Overwriting resources has to be tested excessively, here is a list of all scenarios that have to work:

Regular deployment - app theme:
- [x] css
- [x] js
- [x] images
- [x] templates
- [x] mimetype icons

Subdir deployment - app theme:
- [x] css
- [x] js
- [x] images
- [x] templates
- [x] mimetype icons

Regular deployment - legacy theme:
- [x] css
- [x] js
- [x] images
- [x] templates
- [x] mimetype icons

Subdir deployment - legacy theme:
- [x] css
- [x] js
- [x] images
- [x] templates
- [x] mimetype icons

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Commit:
set themes directory relative from the serverroot and additional the webroot

Commit:
stop unnecessary attempt to look for core templates in the app directory since both and empty string and 'core' should look into the core directory

Commit:
theme directory does not contain trailing or leading slashes, adding them here so templates load properly

Commit:
theme directory does not contain trailing or leading slashes, adding them here so stylesheets load properly

Commit:
theme directory does not contain trailing or leading slashes, adding them here so scripts load properly

Commit:
theme directory does not contain trailing or leading slashes, adding them here so images load properly

Commit:
added missing slash for mimetype icon paths

Commit:
theme directories are now stored without trailing slash

Commit:
removed return type

PVince81 commented on core/js/mimetype.js:
```diff
@@ -82,7 +82,7 @@ OC.MimeType = {
 		// First try to get the correct icon from the current theme
 		if (OC.currentTheme.name !== '' && $.isArray(OC.MimeTypeList.themes[OC.currentTheme.name])) {
-			path = '/' + OC.currentTheme.directory + 'core/img/filetypes/';
+			path = '/' + OC.currentTheme.directory + '/core/img/filetypes/';
```
OC.joinPaths() can help here

PVince81 commented on lib/private/legacy/template.php:
```diff
@@ -189,7 +189,7 @@ public static function initTemplateEngine($renderAs) {
 	protected function findTemplate($theme, $app, $name) {
 		// Check if it is a app template or not.
-		if( $app !== '' ) {
+		if( $app !== '' && $app !== 'core' ) {
```
any reason for this ?

PVince81 replied:
ah right, core is below

phisch replied:
sometimes core templates are loaded with `app=null` and sometimes `app=core` there code for an empty app is slimmer, so it is just a small improvement

PVince81 commented:
:+1:

what's missing ?

Commit:
fixed theme defaults.php path

Commit:
the static method could return false in some scenarios, defaulting to an empty string in that case
this is mainly a concern for automated tests

PVince81 commented:
Stop using type hints for strings or int, that never worked

```
19:12:39 1) Test\Theme\ThemeServiceTest::testSetAppThemeSetsName
19:12:39 Argument 1 passed to OC\Theme\Theme::setWebPath() must be an instance of OC\Theme\string, string given, called in /var/lib/jenkins/workspace/owncloud-core_core_PR-27642-QHFPSK5MZ6Q3ER5MVA5J4VHA5RFTNYSMYBS75F6HQN4LPOS3JAIA/lib/private/Theme/ThemeService.php on line 106 and defined
19:12:39 
19:12:39 /var/lib/jenkins/workspace/owncloud-core_core_PR-27642-QHFPSK5MZ6Q3ER5MVA5J4VHA5RFTNYSMYBS75F6HQN4LPOS3JAIA/lib/private/Theme/Theme.php:90
19:12:39 /var/lib/jenkins/workspace/owncloud-core_core_PR-27642-QHFPSK5MZ6Q3ER5MVA5J4VHA5RFTNYSMYBS75F6HQN4LPOS3JAIA/lib/private/Theme/ThemeService.php:106
19:12:39 /var/lib/jenkins/workspace/owncloud-core_core_PR-27642-QHFPSK5MZ6Q3ER5MVA5J4VHA5RFTNYSMYBS75F6HQN4LPOS3JAIA/tests/lib/Theme/ThemeServiceTest.php:58
```

Commit:
removed typehint

phisch commented:
Had PHPStorm set to PHP7, and used the generate getter/setter feature which then adds type hints for internal data types. Sorry, removed that one and fixed my PHPStorm settings.

PVince81 commented:
Retested, works. Thanks a lot! 👍
